### PR TITLE
Progress bar iteration 2

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Close https://github.com/cowprotocol/cowswap/issues/615, #793 

image
image

[UX Progressbar feedback](https://docs.google.com/document/d/1ZIfWZMu65v-WWCiqeDoLXO272l385Y4_kGS9smP1eaQ/edit)

# To Test

1.  Make a trade
2.  Observe the progress bar

### CODE REVIEW

**Comment by anxolin** [comment]
@alongoni @ramirotw , I think is better if you take this one over. Its best I'm happy to merge this and then you can reiterate in how we present these longer messages. We were going to take a look to this, but theres no capacity from our end right now.

**Comment by ramirotw** [comment]
@alongoni FYI I updated the base branch, cleaned-up and solved conflicts, also did a rebase so if you have issues later with some local changes, ping me

**Review by alfetopito — COMMENTED** [review]
I love this!

Left a few small nitpicks in the code.

Also one small difference between the modal vs activity details

Modal:
Screen Shot 2022-06-24 at 11 09 51
Activity details:
Screen Shot 2022-06-24 at 11 04 10

**Review by elena-zh — CHANGES_REQUESTED** [review]
Hey @alongoni , @ramirotw , great changes!!

However, a list of issues from my side: 
1. Confirmation modal does not appear when place an order in Expert mode!
2. The modal remains to be displayed when an order is expired
expired . It would be nice to hide a progress bar in this case and tell a user that the order is expired. 
3. The same is for Cancel status: a user can cancel an order from the modal, and from another browser (currently, order info is synced between devices). So when an order is cancelling and cancelled, it would be great to close this modal at least. Or it would be great to say a user that order is cancelling/cancelled, and stop showing the progress bar. 
cancelling
cancell
cancelled
4. When connected to Safe wallet, the progress bar is displayed when an order is in Signing state. Progress bar should appear when an order is in 'Open' . Discussed in https://github.com/cowprotocol/cowswap/issues/615#issuecomment-1160339549
safe activity
when signed
safe - usual modal

5. Modal does not look good when screen width is in [724-950] px
Screenshot_4

6. I'm not sire about it, but shouldn't we show different liquidity providers for GC? 
image

7. A bit sophisticated case, but I noticed that when place 2 orders one after another, the conf. and activity modals show different states for progress bar. Firstly, I thought that it shows the 1st order bar (instead of the last), but then I notices that the bar on the conf.modal "lives it own life" and does not match statuses in the activity modal.
diff statuses
diff
I assume, we should show a progress bar for the last order 
8. I noticed, that 'Price out of market' message may be changed to 'Longer than usual' in a while. Is it OK?
9. I noticed that 'Cancel' order links colors are different in the conf. and activity modals. Frakly, it looks good to me, however, I would better clarify
diff colors ok

10. Cancel order button shifted a bit from the main text in a mobile device
image

11. And there is no pop-up saying that a user can track the order execution status when closing the confirmation modal. Will it be implemented in another PR? 

Thanks!

**Comment by elena-zh** [comment]
I have created a separate issue  #745  for the case 11

**Comment by alongoni** [comment]
Thanks for the feedback! I've tackle all styles issues. @ramirotw will continue with the missing points (code improvements and bugs)

File: src/custom/components/AccountDetails/Transaction/OrderProgressBar/index.tsx [hunk 1]
```diff
@@ -7,39 +7,64 @@ import {
...
 export function OrderProgressBar(props: OrderProgressBarProps) {
-  const { activityDerivedState, creationTime, validTo, chainId } = props
-  const { isConfirmed, isCancellable, isUnfillable = false } = activityDerivedState
-  const { elapsedSeconds, expirationInSeconds, isPending } = useGetProgressBarInfo(props)
+  const { activityDerivedState, chainId, hideWhenFinished = false } = props
+  const { order, isConfirmed, isCancellable, isUnfillable = false } = activityDerivedState
+  const { validTo, creationTime } = useMemo(() => {
+    if (order) {
+      return {
+        validTo: new Date((order.validTo as number) * 1000),
```
[the remark this answers is not in the document]
Comment (ramirotw) [reply src/custom/components/AccountDetails/Transaction/OrderProgressBar/index.tsx:51]: it's a Timestamp and 
image

File: src/custom/components/AccountDetails/Transaction/OrderProgressBar/index.tsx [hunk 2]
```diff
@@ -130,20 +196,35 @@ export function OrderProgressBar(props: OrderProgressBarProps) {
...
+                  ) : null}
+                </StatusMsg>
+              </StatusWrapper>
+              <StatusGraph>
+                {/*<p>
+                  Current price: <strong>$1200.56</strong>
+                </p>
+                <p>
+                  Your price: $1300.55 (<span>+8%</span>)
+                </p>*/}
```
[the remark this answers is not in the document]
Comment (ramirotw) [reply src/custom/components/AccountDetails/Transaction/OrderProgressBar/index.tsx:228]: Not really, we should probably add this in another PR. The issue with the current price is that is not saved on the state when fetched to toggle the isUnfillable flag. I didn't want to add more complexity to the PR so I didn't include any changes to handle this.

**Comment by elena-zh** [comment]
Hey @ramirotw , @alongoni great changes. 
I have retested all cases except the 1st one, and Gnosis safe case. 
And here is my feedback: 
1. Progress bar starts is already moving when I sign an order. The it goes back. The time depends on the time I take for signing an order (the request is sent to the wallet, but not yet signed). See the [video](https://watch.screencastify.com/v/YK1JNEhokCYAbreLkNgl). I expect to see the bar from the start when I sign an order. 
2. Coincidence of wants link behaves differently on hover action in the modal and in the activity list
Screenshot_3
Screenshot_6
3. Cancel link is flashing in the activity modal when hover a mouse on it
flash

5. I have not noticed 'Price out of range' message. I placed several orders that should affect a market price, and an opened order hang for 10 minutes with another message, not 'price out of range'
more than 6, no price out of range
6. When an order is 'cancelling', the modal with progress bar still shows that it is executed. Accidentally, the modal might be changed to 'Transaction Submitted' state until the order is cancelled/traded. I think, we should add 'Cancelling' state
cancelling
cancellint

7. Maybe we should not name the modal 'Transaction', as we are placing orders, and orders are displayed on Explorer. WDYT? 
image
image
image

Thanks!

**Comment by ramirotw** [comment]
Thanks @elena-zh for the feedback.

> 1. Progress bar starts is already moving when I sign an order. The it goes back. The time depends on the time I take for signing an order (the request is sent to the wallet, but not yet signed). See the [video](https://watch.screencastify.com/v/YK1JNEhokCYAbreLkNgl). I expect to see the bar from the start when I sign an order.

This is side effect of the fix for:

> A bit sophisticated case, but I noticed that when place 2 orders one after another, the conf. and activity modals show different states for progress bar. Firstly, I thought that it shows the 1st order bar (instead of the last), but then I notices that the bar on the conf.modal "lives it own life" and does not match statuses in the activity modal.

The original problem was that the `creationTime` field from the order is created dynamically and persisted in the storage. If you happen to be using 2 tabs that `creationTime` value has a small difference in seconds when loaded into redux. During the tests I did it was something between 3-5 seconds, small difference but enough for the progress bar percentage to be off. To kind of fix that issue, I first use the `creationTime` as it is a local value, and then when the order details are fetched from the API I use the `creationDate` from the response which is consistent, here it's when you see the jump to the beginning.

**Comment by ramirotw** [comment]
> image

We don't have intermediate on the progress bar, what should we do for a cancelling state? hide it?

**Comment by elena-zh** [comment]
> Thanks @elena-zh for the feedback.
> 
> > 1. Progress bar starts is already moving when I sign an order. The it goes back. The time depends on the time I take for signing an order (the request is sent to the wallet, but not yet signed). See the [video](https://watch.screencastify.com/v/YK1JNEhokCYAbreLkNgl). I expect to see the bar from the start when I sign an order.
> 
> This is side effect of the fix for:
> 
> > A bit sophisticated case, but I noticed that when place 2 orders one after another, the conf. and activity modals show different states for progress bar. Firstly, I thought that it shows the 1st order bar (instead of the last), but then I notices that the bar on the conf.modal "lives it own life" and does not match statuses in the activity modal.
> 
> The original problem was that the `creationTime` field from the order is created dynamically and persisted in the storage. If you happen to be using 2 tabs that `creationTime` value has a small difference in seconds when loaded into redux. During the tests I did it was something between 3-5 seconds, small difference but enough for the progress bar percentage to be off. To kind of fix that issue, I first use the `creationTime` as it is a local value, and then when the order details are fetched from the API I use the `creationDate` from the response which is consistent, here it's when you see the jump to the beginning.

Thanks, @ramirotw . for explanation. But maybe as another option we should not allow the progress bar to move until the data is recorded into redux, WDYT?

**Comment by elena-zh** [comment]
> > image
> 
> We don't have intermediate on the progress bar, what should we do for a cancelling state? hide it?

As an option, we can show something like a message 'Order cancel request is submitted' or so instead of the progress bar.

**Comment by elena-zh** [comment]
Hey @alongoni , I have retested changes in a mobile version, and have noticed that Cancel order button is still shifted
image

It is also shifted in 'longer than usual' message. 
Could you please take a look at it? 

Thanks

File: src/custom/utils/trade.ts [hunk 3]
```diff
@@ -121,6 +120,8 @@ export async function signAndPostOrder(params: PostOrderParams): Promise<AddUnse
     owner: account,
   })
 
+  const creationTime = new Date().toISOString()
```
Comment (ramirotw) [inline src/custom/utils/trade.ts:123]: moved this after the signing is done to reduce the difference with the `creationDate` field from the API if the user takes a few seconds to sign.

**Comment by elena-zh** [comment]
Hey @ramirotw , thank you! The bar is not moving back now, ang the modal name looks great from my perspective.
Cancelling order functionality works fine besides the case that the progress bar modal might be updated with a new status within 30 seconds. Also. I'd remove 'is' from the modal name, as we do not use it for all the rest modal names (Order Cancelled, Order Submitted, Order Expired, etc.)
is

2 more cases I'd like to draw your attention: 
1. Progress bar icon now is starting to move from outside of the bar. Once I caught it even outside the modal area. Could we please fix it somehow?
outside
2. Price out of range message is displayed in the Progress bar modal, but at the same time 'order is taking longer than usual' is displayed in the activity modal for the same order. Can we align these messages, please?
diff statuses

Thank you!

**Comment by ramirotw** [comment]
> 2\. Price out of range message is displayed in the Progress bar modal, but at the same time 'order is taking longer than usual' is displayed in the activity modal for the same order. Can we align these messages, please?
>     diff statuses

That's weird. I imagine this has something to do with which tab has the focus as maybe the state is not updated for that one? could you try to see if this happens between two devices? mobile phone and laptop/desktop?

Regarding the other issue, will take a look.

**Comment by elena-zh** [comment]
@ramirotw , both windows were focused (one after another), as I was traying to do a screenshot that contains both tabs.

**Comment by elena-zh** [comment]
@ramirotw , the issue with different messages is also reproducible when place an order in 1 device, and track it in activity modal in another device
 - mobile phone: 
image

the same wallet connected in desktop:
ьуыы

**Comment by elena-zh** [comment]
Hey @alongoni , links color looks good now!

As for for the Cancel order links, so this is how they look like now: 
image
image

**Comment by elena-zh** [comment]
> @ramirotw , the issue with different messages is also reproducible when place an order in 1 device, and track it in activity modal in another device
> 
> * mobile phone:
>   image
> 
> the same wallet connected in desktop: ьуыы

I have created a separate issue for this case https://github.com/cowprotocol/cowswap/issues/789

**Comment by elena-zh** [comment]
Hey @ramirotw , I'm still testing (need to check other SC wallets), but the issue I have noticed that Progress bar appears and is removed from the confirmation modal when transaction is signed and executed. 
Then, it is blinking when I open Activity modal. 
See the video: https://watch.screencastify.com/v/MWvuTN42VJcg6phYWr3n

Btw, a similar issue I reported for the Cancel button #702 , but the priority was low as 'Cancel' button is not so visible as the progress bar

**Comment by elena-zh** [comment]
Also, still COW icon starts moving from outside progress bar area. 

And cancel button is still blinking when hover a mouse on it https://watch.screencastify.com/v/RxAWdu0wNGVieuS0uklx . Let me know whether I need to create a separate issue for this in case we decide to fix it a bit later.

**Comment by elena-zh** [comment]
> And cancel button is still blinking when hover a mouse on it https://watch.screencastify.com/v/RxAWdu0wNGVieuS0uklx . Let me know whether I need to create a separate issue for this in case we decide to fix it a bit later.

Here is a separate issue for this #793

**Comment by elena-zh** [comment]
I have retested the bar with another SC wallet (Argent), and have not encountered any issues with it (no progress bar there). 

Also, I still see that the progress bar icon starts from outside bar's area. Especially it is visible on mobile devices: 
image

**Comment by alfetopito** [comment]
> > Hey @ramirotw , I'm still testing (need to check other SC wallets), but the issue I have noticed that Progress bar appears and is removed from the confirmation modal when transaction is signed and executed.
> > Then, it is blinking when I open Activity modal.
> > See the video: https://watch.screencastify.com/v/MWvuTN42VJcg6phYWr3n
> 
> The hook that detects if the current wallet is a contract wallet goes through different states before it finally detects wether it is or not. Not sure how to solve this at this point.

Can you start the activity as disabled until you get a concrete answer from the smart contract hook to then decide if it should be shown or continue hidden?

**Comment by ramirotw** [comment]
> Can you start the activity as disabled until you get a concrete answer from the smart contract hook to then decide if it should be shown or continue hidden?

I can't think of a good solution that doesn't imply just delaying showing the progress bar until we are confident the `isSmartContractWallet` value is settled. During my tests the value was settled in less than a second, so maybe 2 seconds will be enough.

I tested it through Wallet Connect and there's no other hint rather than `isSmartContractWallet` when the state changes.
image

**Comment by elena-zh** [comment]
> video: https://watch.screencastify.com/v/MWvuTN42VJcg6phYWr3n

I have updated #702 issue description and added the progress bar issue there. So both that cases will be addresses a bit later in a separate PR.

**Review by elena-zh — APPROVED** [review]
Changes LGTM now!

However, still I can face some issues: 
 - Progress bar can start moving with a delay up to 7 seconds (Mainnet)
 - https://github.com/cowprotocol/cowswap/issues/674#issuecomment-1182852959 - it is a known issue, but now this message has become more visible as the status is displayed on an order confirmation modal now
 - https://github.com/cowprotocol/cowswap/issues/702 - blinking progress bar issue in Gnosis safe (agreed to fix it later)
 - https://github.com/cowprotocol/cowswap/issues/789 - edge case when track an order execution on 2 devices
 - 
 -

**Pull request merged**